src/lib/reasoners/adt_rel.ml

```diff
@@ -55,17 +55,83 @@ type t =
     seen_destr : SE.t;
+    (* Set of all the guarded destructors known by the theory. A guarded
+       destructor isn't interpreted by the ADT theory.
+
+       This field is used to prevent transforming twice a guarded
+       destructor to its non-guarded version and for debugging purposes. *)
+
+    (* TODO: rename this field. *)
     seen_access : SE.t;
+    (* Set of all the non-guarded destructors known by the theory. Contrary
+       to a guarded destructor, a non-guarded destructor is interpreted
+       by the ADT theory. *)
+
     seen_testers : HSS.t MX.t;
     [@ocaml.ppwarning "selectors should be improved. only representatives \
                        in it. No true or false _is"]
+    (* TODO: This code is outdated. It seems that the theory doesn't need
+       to register all the testers in [add_aux]. Now this field is only
+       read to remember we have already forced a constructor because we have
+       assumed its associated tester. *)
 
     selectors : (E.t * Ex.t) list MHs.t MX.t;
+    (* Map of pending destructor equations. A destructor equation on an
+       ADT term [t] is an equation of the form:
+         d t = d' t
+       where [d] is a guarded destructor and [d'] its non-guarded version.
+
+       More precisely, this map matches a representative [r] with
+       all the destructor equations on terms of the class of [r]. These
+       destructor equations are classified by their corresponding
+       constructor.
+
+       Consider [d] a guarded destructor and [c] its associated constructor.
+
+       - When we register the destructor application [d t] in [add_aux], we
+         produce an equation [d t = d' t] where [d'] is the non-guarded
+         version of [d]. If we don't already know that the value of [t] has
+         to use the constructor [c], we put the equation [d t = d' t] in this
+         map. Otherwise we propagate the equation to CC(X).
+
+       - When we assume a tester on [c] (see [assume_is_constr]), we retrieve
+         and propagate to CC(X) all the pending destructor equations
+         associated with the constructor [c].
+
+       - *)
+
     size_splits : Numbers.Q.t;
+    (* Product of the size of all the facts learnt by CS assumed in
+       the theory.
+
+       Currently this field is unused. *)
+
     new_terms : SE.t;
+    (* Set of all the constructor applications built by the theory.
+       See the function [deduce_is_constr]. *)
+
+    (* TODO: rename this field. *)
     pending_deds :
       (r Sig_rel.literal * Ex.t * Th_util.lit_origin) list
+      (* Pending queue of facts discovered by the theory. Facts are pending
+         in the following situations:
+
+         - The domain of a uninterpreted semantic value becomes a singleton.
+           In this case, we deduce the constructor of the semantic value and
+           we add this equation to the pending queue using [deduce_is_constr].
+
+         - When we add a guarded destructor for the first time, we add a new
```

Comment:
Do you mean when the guard is activated (i.e. when we learn the tester)?

Reply:
It's the opposite situation. We know that the domain is a singleton, thus we call `deduce_is_constr`. This function propagate two literals to CC(X) through the queue `pending_deds`:
- an equation of the form `t = cons x_1 ... x_n`
- the literal `(_ cons t)`.
Thus in this situation we activate the guard because the domain is a singleton.

Reply:
If we have already assumed the tester of the associated constructor of the destructor `d` and we add `d` to the theory, we don't create a pending equation in `selectors`.